### PR TITLE
fix(badger): resolve a missing link-count key to the type-based default

### LINKED ISSUE #1973: badger tx GetLinkCount returns (0, nil) for a missing key instead of the type-based default
`badgerTransaction.GetLinkCount` (`pkg/metadata/store/badger/transaction.go:909`) maps a missing `l:` key to `(0, nil)`:

```go
item, err := tx.txn.Get(keyLinkCount(fileID))
if err == badgerdb.ErrKeyNotFound {
    return 0, nil
}
```

`fileLinkCountTxn` (`objects.go:837`) handles the same missing key by returning the type-based default — 2 for a directory, 1 for a regular file. Both are reachable in-transaction on the same rows, so the same file can report nlink 0 or nlink 2 depending on which helper the caller reached for.

A link count of 0 for a live directory is not a benign default. Callers that read-modify-write nlink will persist the wrong value, and 0 reads as "fully unlinked" to anything that treats it as a liveness signal.

This predates #1972 and is out of that PR's scope — surfaced by its reviewer while consolidating the six open-coded link-count lookups onto `fileLinkCountTxn`. Related in spirit to #1897, which fixed the sqlite/postgres variant of "GetLinkCount silently returns a wrong count".

Fix: route the missing-key case through the same type-based default as `fileLinkCountTxn`, or have `GetLinkCount` call it. Worth checking the sqlite and postgres transaction twins for the same divergence at the same time — both were noted during the audit as returning `(0, nil)`.

### LINKED ISSUE #1980: badger updateWithConflictRetry returns an unclassified conflict on retry exhaustion
`pkg/metadata/store/badger/` has two retry loops that its own comments describe as the same thing, and they diverge on what they return when retries run out.

`withTransaction` (`transaction.go:184-212`) classifies the exhausted SSI abort — it wraps `badgerdb.ErrConflict` through `mapBadgerError` into `StoreError{Code: ErrConflict}` and bumps `s.txnConflicts`. Its comment states plainly why: so that codebase-wide conflict detection (`errors.As(*StoreError)` / `IsConflictError`, `mapObjectIDConflict`, `isObjectIDConflict`) recognises it.

`updateWithConflictRetry` (`objects.go:157-174`) does neither. It returns the bare `badgerdb.ErrConflict` sentinel and counts nothing.

`IsConflictError` (`pkg/metadata/errors/errors.go:394`) only matches a `*StoreError` with `ErrConflict`:

```go
func IsConflictError(err error) bool {
    var storeErr *StoreError
    if goerrors.As(err, &storeErr) {
        return storeErr.Code == ErrConflict
    }
    return false
}
```

A bare sentinel never matches. So on badger, an exhausted conflict escapes as an unclassified error while postgres and sqlite classify the same condition — callers that retry or degrade on a conflict instead see an opaque failure, and `txnConflicts` under-reports, which is the metric tests assert on to prove a workload stayed conflict-free.

**Affected callers:** `IncrementRefCount`, `DecrementRefCount`, `AddRef`, the reap path and `DecrLiveChunkCount` (`objects.go:157/182/210/245/301`), plus `block_record_store.go:280`. #1972 adds the share lifecycle ops — `CreateShare`, `UpdateShareOptions`, `DeleteShare`, `CreateRootDirectory` — so the blast radius grows with that PR.

Bounded to the post-retry tail, which is why it is LOW rather than higher: it needs `maxTransactionRetries` consecutive conflicts to reach.

Fix is about five lines — mirror `withTransaction`'s tail: bump `s.txnConflicts` on each conflict and map `lastErr` before returning. Better still, have both loops share one helper so the drift cannot recur; that is the structural half and belongs with the other badger-internal duplication in #1828.

Found by the LOW `structure` triage of the 2026-08-05 data-plane audit — a "these two are duplicated" finding whose two copies had silently drifted.

### PR DESCRIPTION
**Pull request by marmos91**
Closes #1973.

**Scope note:** this branch was cut to close #1973 *and* #1980. While it was in review, #1972 merged and fixed #1980 completely — `updateWithConflictRetry` now bumps `txnConflicts` and maps the exhausted sentinel through `mapBadgerError`, with `conflict_classification_test.go` covering both the exhausted and the non-conflict-passthrough cases. That commit has been dropped from this branch; #1980 can be closed against #1972. The one leftover is the second commit here: `txnConflicts`' field comment still named only `withTransaction` after both loops started bumping it.

---

## `GetLinkCount` returned 0 for a live inode

`badgerTransaction.GetLinkCount` mapped a missing `l:<uuid>` key to `(0, nil)`. Every other surface in the package resolves that same absent key to the type-based default — 2 for a directory, 1 otherwise — and after #1972's consolidation they all route through `fileLinkCountTxn`. `GetLinkCount` was the last call site left open-coding the wrong answer, so one inode reported nlink 0 or nlink 2 depending on which helper the caller reached for. 0 is not a benign default: it reads as "fully unlinked" to anything treating the count as a liveness signal.

**The `(0, nil)` was itself the bug at the call sites**, not just an inconsistency:

- `directory.go` rmdir guards the parent decrement on `parentLinkCount > 0`, so a 0 silently skipped the parent's nlink decrement entirely.
- `file_create.go` mkdir and `file_modify.go` rename both do `SetLinkCount(dir, count+1)`, persisting 1 where the parent should have gone to 3.

The fix routes the missing-key case through `fileLinkCountTxn`, which already owns that default, so the two cannot drift again. A missing `f:` record still resolves to 0 — no inode, no links — matching sqlite and postgres. A real read error is propagated rather than fabricated as 0.

**Reachability.** No resurrected state: `deleteFileKeys` removes `f:` and `l:` together, and the last-link unlink in `file_remove.go` *writes* `l: = 0` rather than deleting it, so a genuinely unlinked file never enters the new branch. It is reached only by an inode whose `f:` record exists but whose `l:` key was never written — exactly the state the other overlays in this package already special-case.

## Contract

The conformance suite never covered this: `createTestDir` / `createTestFile` always call `SetLinkCount`, so "PutFile without SetLinkCount" was untested on every backend. Added `DirOps/LinkCountAgreesWithGetFile` to `pkg/metadata/storetest/`, asserting `GetLinkCount == GetFile().Nlink` and non-zero for such an inode. It deliberately pins agreement rather than a fixed number, since the SQL backends land on 1 via the `nlink INTEGER NOT NULL DEFAULT 1` column default while badger derives 2 for a directory.

Before the fix:

```
--- FAIL: TestConformance/DirOps/LinkCountAgreesWithGetFile/Directory
--- FAIL: TestConformance/DirOps/LinkCountAgreesWithGetFile/RegularFile
ok  	pkg/metadata/store/memory
ok  	pkg/metadata/store/sqlite
```

Memory and sqlite already satisfied it; badger was the only backend that did not. Postgres shares sqlite's DDL default and an identical inode INSERT that omits `nlink`, so it passes for the same reason (CI covers it under the integration tag).

## Verification

`go build ./...`, `go vet ./pkg/...`, `gofmt`, `golangci-lint run ./pkg/metadata/...` (0 issues), `go test ./pkg/metadata/... ./pkg/block/...`, and `go test -race` on the badger conformance tests. Protocol suites not run — the change touches no wire behaviour.

### CODE REVIEW

**Comment by marmos91** [comment]
Force-pushed: #1972 merged mid-review and closed #1980 in full (`updateWithConflictRetry` now counts and maps the exhausted conflict, with `conflict_classification_test.go` covering it). That commit is dropped here and #1980 should be closed against #1972. What remains is the #1973 fix, rebased onto the consolidated `fileLinkCountTxn`, plus a one-line correction to the now-stale `txnConflicts` field comment.

Also addressed the review comment on the conformance test: it now sets the parent edge alongside the child edge, so the only thing left unset is the link count.

File: pkg/metadata/storetest/dir_ops.go [hunk 1]
```diff
@@ -293,3 +294,76 @@ func testRootDirectoryIdempotent(t *testing.T, factory StoreFactory) {
...
+					GID:  1000,
+				},
+			}
+			entry.ID = id
+			if err := store.PutFile(ctx, entry); err != nil {
+				t.Fatalf("PutFile() failed: %v", err)
+			}
+			if err := store.SetChild(ctx, rootHandle, name, handle); err != nil {
+				t.Fatalf("SetChild() failed: %v", err)
+			}
```
[the remark this answers is not in the document]
Comment (marmos91) [reply pkg/metadata/storetest/dir_ops.go:356]: Agreed, done. The test now calls SetParent alongside SetChild, matching createTestDir, so the missing link count is the only variable left unset.
